Mathlib/RingTheory/Valuation/ValuativeRel.lean

```diff
@@ -0,0 +1,587 @@
+
+end Valuation
+
+/-- A preorder on a ring is said to be "valuative" if it agrees with the
+valuative relation. -/
+class ValuativePreorder (R : Type*) [CommRing R] [ValuativeRel R] [Preorder R] where
+  rel_iff_le (x y : R) : x ≤ᵥ y ↔ x ≤ y
+
+namespace ValuativeRel
+
+variable {R : Type*} [CommRing R] [ValuativeRel R]
+
+@[simp]
+lemma rel_refl (x : R) : x ≤ᵥ x := by
+  cases rel_total x x <;> assumption
+
+lemma rel_rfl {x : R} : x ≤ᵥ x :=
+  rel_refl x
+
+protected alias rel.refl := rel_refl
+
+protected alias rel.rfl := rel_rfl
+
+@[simp]
+theorem zero_rel (x : R) : 0 ≤ᵥ x := by
+  simpa using rel_mul_right x ((rel_total 0 1).resolve_right not_rel_one_zero)
+
+lemma rel_mul_left {x y : R} (z) : x ≤ᵥ y → (z * x) ≤ᵥ (z * y) := by
+  rw [mul_comm z x, mul_comm z y]
+  apply rel_mul_right
+
+instance : Trans (rel (R := R)) (rel (R := R)) (rel (R := R)) where
+  trans h1 h2 := rel_trans h1 h2
+
+lemma rel_mul {x x' y y' : R} (h1 : x ≤ᵥ y) (h2 : x' ≤ᵥ y') : x * x' ≤ᵥ y * y' := by
+  calc x * x' ≤ᵥ x * y' := rel_mul_left _ h2
+    _ ≤ᵥ y * y' := rel_mul_right _ h1
+
+theorem rel_add_cases (x y : R) : x + y ≤ᵥ x ∨ x + y ≤ᵥ y :=
+  (rel_total y x).imp (fun h => rel_add .rfl h) (fun h => rel_add h .rfl)
+
+variable (R) in
+/-- The submonoid of elements `x : R` whose valuation is positive. -/
+def posSubmonoid : Submonoid R where
+  carrier := { x | ¬ x ≤ᵥ 0}
+  mul_mem' {x y} hx hy := by
+    dsimp only [Set.mem_setOf_eq] at hx hy ⊢
+    contrapose! hy
+    rw [show (0 : R) = x * 0 by simp, mul_comm x y, mul_comm x 0] at hy
+    exact rel_mul_cancel hx hy
+  one_mem' := not_rel_one_zero
+
+@[simp]
+lemma posSubmonoid_def (x : R) : x ∈ posSubmonoid R ↔ ¬ x ≤ᵥ 0 := Iff.refl _
+
+@[simp]
+lemma right_cancel_posSubmonoid (x y : R) (u : posSubmonoid R) :
+    x * u ≤ᵥ y * u ↔ x ≤ᵥ y := ⟨rel_mul_cancel u.prop, rel_mul_right _⟩
+
+@[simp]
+lemma left_cancel_posSubmonoid (x y : R) (u : posSubmonoid R) :
+    u * x ≤ᵥ u * y ↔ x ≤ᵥ y := by
+  simp only [← right_cancel_posSubmonoid x y u, mul_comm]
+
+variable (R) in
+/-- The setoid used to construct `ValGroupWithZero R`. -/
+def valueSetoid : Setoid (R × posSubmonoid R) where
+  r := fun (x,s) (y,t) => x * t ≤ᵥ y * s ∧ y * s ≤ᵥ x * t
+  iseqv := {
+    refl ru := ⟨rel_refl _, rel_refl _⟩
+    symm h := ⟨h.2, h.1⟩
+    trans := by
+      rintro ⟨r, u⟩ ⟨s, v⟩ ⟨t, w⟩ ⟨h1, h2⟩ ⟨h3, h4⟩
+      constructor
+      · have := rel_mul h1 (rel_refl ↑w)
+        rw [mul_right_comm s] at this
+        have := rel_trans this (rel_mul h3 (rel_refl _))
+        rw [mul_right_comm r, mul_right_comm t] at this
+        simpa using this
+      · have := rel_mul h4 (rel_refl ↑u)
+        rw [mul_right_comm s] at this
+        have := rel_trans this (rel_mul h2 (rel_refl _))
+        rw [mul_right_comm t, mul_right_comm r] at this
+        simpa using this
+  }
+
+variable (R) in
+/-- The "canonical" value group-with-zero of a ring with a valuative relation. -/
+def ValGroupWithZero := Quotient (valueSetoid R)
```

Comment:
I am personally very much on the "I don't care how long it is, I want to stick to the literature as much as possible rather than make mathlibisms which only we understand" (an egregious example I saw recently was `Doset` which is presumably justified by "DoubleCoset is too long" but the counterargument is "yes but Double Coset is also *what they are called in the books*!"). The books say "value group", this is a variant which is the value group with an added zero, so I am absolutely for "ValueGroupWithZero" here. I think that the cost of dropping two letters to cause obscurity outweighs the benefits that we win by brevity. But I know other people have other opinions (e.g. I think Yael values brevity at all costs and hang what the books say).